File: ui-preview/src/main/java/l/files/thumbnail/MediaThumbnailer.java
package l.files.thumbnail;

import android.content.Context;
import android.media.MediaMetadataRetriever;

import l.files.fs.Path;
import l.files.ui.base.graphics.Rect;
import l.files.ui.base.graphics.ScaledBitmap;

import static l.files.ui.base.media.MediaMetadataRetrievers.getAnyThumbnail;

final class MediaThumbnailer implements Thumbnailer<Path> {

    @Override
    public boolean accepts(Path path, String mediaType) {
        return mediaType.startsWith("audio/") ||
                mediaType.startsWith("video/");
    }

    @Override
    public ScaledBitmap create(Path path, Rect max, Context context)
            throws Exception {
        MediaMetadataRetriever retriever = new MediaMetadataRetriever();
        try {
            retriever.setDataSource(context, path.toUri());
            return getAnyThumbnail(retriever, max);
        } finally {
            retriever.release();
        }
    }

}

File: ui-preview/src/main/java/l/files/thumbnail/MediaThumbnailer.kt
package l.files.thumbnail

import android.content.Context
import android.media.MediaMetadataRetriever
import l.files.ui.base.graphics.Rect
import l.files.ui.base.media.MediaMetadataRetrievers
import java.nio.file.Path

internal object MediaThumbnailer : Thumbnailer<Path> {

  override fun accepts(path: Path, type: String) =
    type.startsWith("audio/") ||
      type.startsWith("video/")

  override fun create(input: Path, max: Rect, context: Context) =
    MediaMetadataRetriever().let { retriever ->
      try {
        retriever.setDataSource(input.toString())
        MediaMetadataRetrievers.getAnyThumbnail(retriever, max)
      } finally {
        retriever.release()
      }
    }
}
